src/sys/freebsd.rs

```diff
@@ -1 +1,268 @@
+        XAttrs {
+            user_attrs: Vec::from(&*self.user_attrs).into_boxed_slice(),
+            system_attrs: Vec::from(&*self.system_attrs).into_boxed_slice(),
+            offset: self.offset
+        }
+    }
+
+    fn clone_from(&mut self, other: &XAttrs) {
+        self.offset = other.offset;
+
+        let mut data = mem::replace(&mut self.user_attrs, Box::new([])).into_vec();
+        data.extend(other.user_attrs.iter().cloned());
+        self.user_attrs = data.into_boxed_slice();
+
+        data = mem::replace(&mut self.system_attrs, Box::new([])).into_vec();
+        data.extend(other.system_attrs.iter().cloned());
+        self.system_attrs = data.into_boxed_slice();
+    }
+}
+
+impl Iterator for XAttrs {
+    type Item = OsString;
+    fn next(&mut self) -> Option<OsString> {
+        if self.user_attrs.is_empty() && self.system_attrs.is_empty() {
+            return None;
+        }
+
+        if self.offset == self.user_attrs.len() + self.system_attrs.len() {
+            return None;
+        }
+
+        let data = if self.offset < self.system_attrs.len() {
+            &self.system_attrs[self.offset..]
+        } else {
+            &self.user_attrs[self.offset - self.system_attrs.len() ..]
+        };
+
+        let siz = data[0] as usize;
+
+        self.offset += siz + 1;
+        if self.offset < self.system_attrs.len() {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_SYSTEM))
+        } else {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_USER))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        if self.user_attrs.len() + self.system_attrs.len() == self.offset {
+            (0, Some(0))
+        } else {
+            (1, None)
+        }
+    }
+}
+
+fn name_to_ns(name: &OsStr) -> io::Result<(c_int, CString)> {
+    let name_str = match name.to_str() {
+        Some(s) => s,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "cannot convert to rust string"))
+    };
+
+    let dot = name_str.find('.').unwrap();
+    let (nsname, propname) = name_str.split_at(dot);
+    let ns_int = match EXTATTR_NAMESPACE_NAMES.iter().position(|&s| s == nsname) {
+        Some(i) => i,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "no matching namespace"))
+    };
+
+    Ok((ns_int as c_int, try!(name_to_c(OsStr::new(&propname[1..])))))
+}
+
+fn prefix_namespace(attr: OsString, ns: c_int) -> OsString {
+    let nsname = EXTATTR_NAMESPACE_NAMES[ns as usize];
+    let mut vattr = attr.into_vec();
+    let attr_len = vattr.len();
+    let nsname_len = nsname.len();
+
+    vattr.reserve(nsname_len + 1);
+    unsafe {
+        vattr.set_len(nsname_len + attr_len + 1);
+        ptr::copy(vattr.as_ptr(), vattr[nsname_len + 1..].as_mut_ptr(), attr_len);
+        ptr::copy(nsname.as_ptr(), vattr.as_mut_ptr(), nsname_len);
+    };
```

Comment:
You don't need this unsafe. Just take `attr` as an `&OsStr` and then build the full string by concatenating the prefix and `attr`.